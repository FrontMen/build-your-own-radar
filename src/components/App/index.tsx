import React from 'react';
import { Router } from '../../Router';

export const App: React.FC = () => (
  <div className="App">
    <Router data-testid="router" />
  </div>
);
