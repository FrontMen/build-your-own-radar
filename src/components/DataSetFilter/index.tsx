import React, { useCallback, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { selectedDataSetSelector } from 'redux/selectors/filters';
import { filtersActions } from 'redux/actions/filters';
import {
  Container,
  Selected,
  Text,
  Triangle,
  Option,
  DropDown,
} from './styled';

export const DataSetFilter = () => {
  const [opened, setOpened] = useState<boolean>(false);
  const { availableDates, selected } = useSelector(selectedDataSetSelector);
  const dispatch = useDispatch();
  const selectDate = useCallback(
    (date: string) => {
      dispatch(filtersActions.selectDataSet(date));
    },
    [dispatch],
  );
  const toggleOpened = useCallback(() => {
    setOpened(prevOpened => !prevOpened);
  }, [setOpened]);

  return (
    <Container data-testid="dataSetFilter-container">
      <Selected data-testid="dataSetFilter-anchor" onClick={toggleOpened}>
        <Text data-testid="dataSetFilter-text">{selected}</Text>
        <Triangle
          opened={opened}
          focusable="false"
          viewBox="0 0 24 24"
          aria-hidden="true"
          role="presentation"
        >
          <path d="M7 10l5 5 5-5z" />
        </Triangle>
      </Selected>
      {opened && (
        <DropDown opened={opened} data-testid="dataSetFilter-dropdown">
          {availableDates.map(date => (
            <Option
              key={date}
              data-testid={`dataSetFilter-dropdown-option-${date}`}
              selected={selected === date}
              onClick={() => {
                selectDate(date);
                setOpened(false);
              }}
            >
              {date}
            </Option>
          ))}
        </DropDown>
      )}
    </Container>
  );
};
