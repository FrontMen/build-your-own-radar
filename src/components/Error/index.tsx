import React from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import { MainContentSlot } from 'components/shared/PageSlots';
import { auth } from 'utils/auth';

const Content = styled.div`
  display: flex;
  height: 50vh;
  justify-content: center;
  align-items: center;
`;

export interface IError {
  message: string;
}

export const Error: React.FC<IError> = ({ message }) => {
  const history = useHistory();
  if (message.includes('token') || !auth.isLoggedIn()) {
    auth.logout();
    history.push('/auth/login');
    return null;
  }
  return (
    <MainContentSlot>
      <Content data-testid="home-error">
        <div>
          <h1>500</h1>
          <h2>Oops! Looks like the server is not responding</h2>
          <h3>{message}</h3>
        </div>
      </Content>
    </MainContentSlot>
  );
};
