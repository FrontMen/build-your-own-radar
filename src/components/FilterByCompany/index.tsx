import React, { useContext } from 'react';
import {  CheckBox } from 'src/components/shared/CheckBox';
import styled from 'styled-components';
import { filterByCompanyContext } from 'src/ContextProviders/FilterByCompanyContextProvider';

const COMPANY_NAMES: {
  [K in CompanyTypes]: string;
} = {
  ITR_BE: 'Intracto BE',
  ITR_NL: 'Intracto NL',
  FM: 'Frontmen',
};

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: ${props => props.theme.space[2]}px 0;
`;

export const FilterByCompany = () => {
  const { state, toggle } = useContext(filterByCompanyContext);
  return (
    <Container>
      {Object.entries(state).map(([company, checked]) => (
        <CheckBox
          key={company}
          label={COMPANY_NAMES[company as CompanyTypes]}
          checked={checked}
          onClick={() => toggle(company as CompanyTypes)}
        />
      ))}
    </Container>
  );
};
