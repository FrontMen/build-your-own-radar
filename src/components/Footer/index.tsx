import React from 'react';
import styled from 'styled-components';
import { FooterSlot } from '../shared/PageSlots';
import { MediaQueries } from 'Theme/Helpers';
import facebookLogo from 'res/svg/facebook.svg';
import instagramLogo from 'res/svg/instagram.svg';
import youtubeLogo from 'res/svg/youtube.svg';
import linkedinLogo from 'res/svg/linkedin.svg';

const Slot = styled(FooterSlot)`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const Links = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
  flex-wrap: wrap;
  border-bottom: 1px solid ${props => props.theme.pallet.grayLight};
  padding-bottom: 12px;
`;

const IntractoLinks = styled.div`
  a {
    display: inline-block;
    width: 30px;
    height: 30px;
    margin-right: 15px;
    opacity: 0.7;

    &:hover {
      opacity: 1;
    }
  }
`;

const FrontmenLink = styled.a`
  align-self: center;
  font-family: Montserrat, san-serif;
  font-weight: 700;
  background-image: url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 80 106'%3E%3Cpath fill='%23201E33' d='M0 0h80v60H0z'/%3E%3Cpath fill='%231F1D32' d='M40.07 0L0 23.16v59.674L39.97 106 80 82.676V23.16z'/%3E%3Cpath fill='%23FFF' d='M10 29v48.49L19.333 83V65.44L38 76.462v-11.02L19.333 54.42v-8.919L38 56.551v-11.02z'/%3E%3Cpath fill='%23FF5900' d='M42 45.53v11.021l18.667-11.02V83L70 77.49V29z'/%3E%3Cpath fill='%230CC' d='M14 26.473L23.464 32 40.5 22.053 57.536 32 67 26.473 40.5 11z'/%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-size: contain;
  background-position-x: right;
  color: ${props => props.theme.pallet.dark};
  text-decoration: none;
  opacity: 0.8;
  margin-top: 10px;

  &::before {
    content: 'In Collaboration with Frontmen';
    margin-right: 20px;
  }

  &:hover {
    opacity: 1;
  }

  @media (${MediaQueries.phablet}) {
    margin-top: 0;
  }
`;

const PrivacyLinks = styled.div`
  margin-top: 12px;

  a {
    margin-right: 10px;
    font-size: 13px;
    text-decoration: none;
    opacity: 0.7;
    color: ${props => props.theme.pallet.grayDark};

    &:hover {
      opacity: 1;
    }
  }
`;

export const Footer = () => (
  <Slot>
    <Links>
      <IntractoLinks>
        <a
          href="https://www.facebook.com/Intracto"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src={facebookLogo} alt="facebook-logo" />
        </a>
        <a
          href="https://www.linkedin.com/company/intracto/"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src={linkedinLogo} alt="linkedin-logo" />
        </a>
        <a
          href="https://www.youtube.com/channel/UCxdOVsGyzWkQ2zxx8ojSxHQ"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src={youtubeLogo} alt="youtube-logo" />
        </a>
        <a
          href="https://www.instagram.com/intracto_agency/"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src={instagramLogo} alt="instagram-logo" />
        </a>
      </IntractoLinks>
      <FrontmenLink
        href="https://www.frontmen.nl/"
        data-testid="footer-Frontmen-link"
        target="_blank"
      />
    </Links>
    <PrivacyLinks>
      <a
        href="https://www.intracto.com/en-be/cookie-policy"
        target="_blank"
        rel="noopener noreferrer"
      >
        cookie policy
      </a>
      <a
        href="https://www.intracto.com/en-be/privacy-policy"
        target="_blank"
        rel="noopener noreferrer"
        data-testid="footer-Intracto-link"
      >
        privacy policy
      </a>
      <a
        href="https://www.intracto.com/en-be/conditions"
        target="_blank"
        rel="noopener noreferrer"
      >
        terms & conditions
      </a>
    </PrivacyLinks>
  </Slot>
);
