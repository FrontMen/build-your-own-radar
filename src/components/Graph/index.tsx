import React, { useCallback, useEffect, useRef } from 'react';
import { useMediaQuery } from 'react-responsive';
import { radar_visualization, showBubble, hideBubble } from 'src/utils/d3';
import { d3Config } from 'src/utils/d3-config';
import styled from 'styled-components';
import { MediaQueries } from 'src/Theme/Helpers';
import { useHistory } from 'react-router';

const GraphWrapper = styled.div<{ fullSize: boolean | undefined }>`
  width: 100%;
  max-width: ${props => (props.fullSize ? '800px' : '440px')};
  min-width: 280px;
  height: auto;
  margin: 0 auto ${props => props.theme.space[2]}px;

  @media ${MediaQueries.phablet} {
    max-width: 80%;
    margin-bottom: ${props => props.theme.space[3]}px;
  }
  @media ${MediaQueries.desktop} {
    max-width: 50%;
    margin-bottom: ${props => props.theme.space[5]}px;
  }
`;

const GraphContainer = styled.div`
  text-align: center;
  @media ${MediaQueries.desktop} {
    position: sticky;
    top: 0;
  }
`;

interface TechnologiesListProps {
  technologies: Technology[];
  highlighted: string | null;
  setHighlighted?: (a: string | null) => void;
  setSelected?: (a: string | null) => void;
  quadrant?: number;
  className?: string | undefined;
  fullSize?: boolean;
}

export const Graph: React.FC<TechnologiesListProps> = ({
  highlighted,
  quadrant,
  technologies,
  setHighlighted = () => {},
  setSelected = () => {},
  className: StylesFromParent,
  fullSize,
}) => {
  const d3Container = useRef<SVGSVGElement>(null);
  const history = useHistory();

  const redirect = useCallback(
    (path: string) => {
      history.push(path);
    },
    [history],
  );

  const isNotMobile = useMediaQuery({ query: MediaQueries.phablet });
  useEffect(() => {
    if (d3Container.current) {
      radar_visualization(
        d3Container.current,
        technologies,
        d3Config,
        setHighlighted,
        setSelected,
        {
          quadrant,
          isNotMobile,
        },
        redirect,
      );
    }
  }, [technologies, quadrant, setHighlighted, setSelected, redirect, isNotMobile]);

  useEffect(() => {
    const technology = technologies?.find(t => t.name === highlighted);
    if (technology) {
      showBubble(technology);
    } else {
      hideBubble();
    }
  }, [highlighted, technologies]);

  return (
    <GraphWrapper fullSize={fullSize}>
      <GraphContainer>
        <svg ref={d3Container} />
      </GraphContainer>
    </GraphWrapper>
  );
};
