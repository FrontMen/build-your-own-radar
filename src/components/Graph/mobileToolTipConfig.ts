export default [
  {
    rotateZ: '-35deg',
    right: '-10%',
    bottom: '-3%',
  },
  {
    rotateZ: '35deg',
    bottom: '5%',
    left: '-15%',
  },
  {
    top: '0',
    left: '-15%',
    rotateZ: '-34deg',
    translate: '-90px, -290px',
  },
  {
    top: '5%',
    right: '-15%',
    rotateZ: '36deg',
    translate: '-90px, -150px',
  },
];
