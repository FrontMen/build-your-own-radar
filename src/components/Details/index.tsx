import React from 'react';
import { Redirect, useParams } from 'react-router-dom';
import styled from 'styled-components';
import { MainContentSlot } from 'components/shared/PageSlots';
import { ContentTitle } from 'components/shared/ContentTitle';
import { DetailsSkeleton } from 'components/Skeleton/Details';
import { Text } from 'components/Text';
import { Link } from 'react-router-dom';
import { Typography } from 'Theme/Typography';
import { IoIosArrowRoundBack } from 'react-icons/io';
import { useSelector } from 'react-redux';
import {
  allTechnologyDataSetSelector,
  technologiesLoadingStateSelector,
} from 'redux/selectors/technologies';
import { quadrantsSelector } from 'redux/selectors/filters';
import { dateFormat, transMapper } from 'utils';

const Slot = styled(MainContentSlot)``;
const BackLink = styled(({ quadName, ...props }) => <Link {...props} />)`
  display: inline-flex;
  align-items: center;
  padding: ${props => `${props.theme.space[2]}px ${props.theme.space[3]}px`};
  padding-left: 0;
  ${Typography.body};
  color: ${props => props.theme.colors[props.quadName]};
  text-decoration: none;
`;

const ArrowLeftIcon = styled(IoIosArrowRoundBack)`
  height: 2em;
  width: 2em;
`;

const Content = styled.div`
  margin: ${props => `${props.theme.space[3]}px 0 ${props.theme.space[4]}px`};
  text-align: center;
`;

const Record = styled.ul<{ color: string }>`
  border-left: 4px solid ${props => props.theme.colors[props.color || '']};
  margin: 50px 0% 0 30%;
  padding: 50px 0 50px 20px;
  list-style: none;
  text-align: left;
  h3 {
    font-weight: 400;
    font-size: 1.4em;
  }
`;

const RecordItem = styled.li<{ color: string; value: string }>`
  border-bottom: 1px dashed ${props => props.theme.colors[props.color || '']};
  padding-bottom: 25px;
  position: relative;
  &:last-of-type {
    padding-bottom: 0;
    margin-bottom: 0;
    border: none;
  }
  &:before,
  &:after {
    position: absolute;
    display: block;
    top: 0;
  }
  &:before {
    left: -160px;
    top: 2px;
    content: '${props => props.value}';
    text-align: right;
    min-width: 120px;
    font-weight: bold;
    font-size: 1.1em;
    letter-spacing: 1px;
  }
  &:after {
    box-shadow: ${props =>
      `0 0 0 4px ${props.theme.colors[props.color || '']}`};
    left: -28px;
    background: #fff;
    border-radius: 50%;
    height: 11px;
    width: 11px;
    content: '';
    top: 5px;
  }
  h3 {
    font-weight: 400;
    font-size: 1.4em;
  }
`;

const Tag = styled.div`
  background: ${props => props.theme.pallet.light};
  max-width: 205px;
  padding: 10px;
  color: ${props => props.theme.pallet.dark};
  border: 1px solid ${props => props.theme.pallet.light};
  border-radius: 6px;
  word-break: break-word;
  font-family: Montserrat, san-serif;
  font-size: 0.9em;
  line-height: 20px;
`;

export interface DetailsParams {
  technology?: string;
  quadIndex?: string;
}

export const Details: React.FC = () => {
  const { quadIndex, technology: technologyParam } = useParams<DetailsParams>();
  const allData = useSelector(allTechnologyDataSetSelector);
  const { loading, initialized } = useSelector(
    technologiesLoadingStateSelector(),
  );
  const quadrants = useSelector(quadrantsSelector);
  const decodedTechName = decodeURIComponent(technologyParam || '');
  const numericQuadIndex = Number(quadIndex);
  const foundQuadrant = quadrants.find(
    item => item.order === Number(numericQuadIndex),
  );

  let lastDescription: string;
  const technologies = Object.entries(allData)
    .sort((a, b) => Date.parse(b[0]) - Date.parse(a[0]))
    .map(([date, data]) => {
      const foundItem = data.find(
        (item: Technology) => item.name === decodedTechName,
      );
      if (foundItem) {
        if (lastDescription !== foundItem.description) {
          lastDescription = foundItem.description;
          return { ...foundItem, date };
        }
      }
      return null;
    })
    .filter(Boolean);

  if (loading || !initialized || quadrants.length === 0)
    return <DetailsSkeleton />;

  if (!foundQuadrant || !technologies.length)
    return <Redirect to="/not-found" />;

  const { name: quadrantName } = foundQuadrant;
  return (
    <Slot>
      <BackLink
        quadName={quadrantName}
        to={`/quadrant/${numericQuadIndex}`}
        data-testid="details-back-link"
      >
        <ArrowLeftIcon />
        <Text value={`quadrant.${transMapper[numericQuadIndex]}.name`} />
      </BackLink>
      <Content>
        <ContentTitle data-testid="details-content-title">{`Record: ${decodedTechName}`}</ContentTitle>
        <Record data-testid="details-record-container" color={quadrantName}>
          {technologies.map((item, index) => (
            <RecordItem
              data-testid={`details-record-item-${index}`}
              key={index}
              value={dateFormat(item!.date)}
              color={quadrantName}
            >
              <h3>{item!.name}</h3>
              {item?.description ? (
                <p>{item!.description}</p>
              ) : (
                <Tag>
                  <Text value="details.emptyDescription" />
                </Tag>
              )}
            </RecordItem>
          ))}
        </Record>
      </Content>
    </Slot>
  );
};
