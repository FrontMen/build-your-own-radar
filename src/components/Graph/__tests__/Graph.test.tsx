import React from 'react';
import { Graph, GraphProps } from 'components/Graph';
import { withAllProviders, AllProvidersWrapper } from 'test/helpers';
import { blips } from 'test/mockData';
import { mount } from 'enzyme';
import { d3Config } from 'utils/d3-config';
import { createMemoryHistory } from 'history';

const defaultProps = {
  blips,
  highlighted: null,
  setHighlighted: jest.fn(),
  setSelected: jest.fn(),
  quadrantNum: 1,
};
jest.mock('utils/d3', () => {
  const actual = jest.requireActual('utils/d3');
  return {
    ...actual,
    radar_visualization: jest.fn(),
    showBubble: jest.fn(),
    hideBubble: jest.fn(),
  };
});

const defer = async () => new Promise(r => setTimeout(r, 0));

describe('Graph', () => {
  /* test to check that the mobile and larger snapshot don't change unintentionally */
  it('should matches the snapshot', () => {
    const { container } = withAllProviders(<Graph {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should call showBubble when technology selected and hideBubble when not', async () => {
    const { showBubble, hideBubble } = await import('utils/d3');
    const history = createMemoryHistory({ initialEntries: ['/'] });
    const wrapper = mount<GraphProps>(<Graph {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
      wrappingComponentProps: {
        history,
      },
    });

    wrapper.setProps({
      highlighted: blips[0].positionId!,
    });
    expect(showBubble).toHaveBeenCalledWith(blips[0], 1);

    wrapper.setProps({
      highlighted: null,
    });
    expect(hideBubble).toHaveBeenCalled();
  });

  it('should call radar_visualization with proper params when technologies or quadrant changes', async () => {
    const history = createMemoryHistory({ initialEntries: ['/'] });
    const {
      radar_visualization,
    }: {
      radar_visualization: jest.MockedFunction<any>;
    } = await import('utils/d3');
    const wrapper = mount<GraphProps>(<Graph {...defaultProps} blips={[]} />, {
      wrappingComponent: AllProvidersWrapper,
      wrappingComponentProps: {
        history,
      },
    });

    wrapper.setProps({
      blips,
    });

    await defer();
    const [_, b] = radar_visualization.mock.calls[1];
    expect(b).toBe(blips);

    radar_visualization.mockClear();

    wrapper.setProps({
      quadrantNum: 3,
    });

    await defer();
    const [
      container,
      bl,
      changed,
      config,
      setHighlighted,
      setSelected,
      setHoveredQuadrant,
      { quadrantNum },
    ] = radar_visualization.mock.calls[0];
    expect(config).toBe(d3Config);
    expect(setHighlighted).toBe(defaultProps.setHighlighted);
    expect(quadrantNum).toBe(3);
  });
});
