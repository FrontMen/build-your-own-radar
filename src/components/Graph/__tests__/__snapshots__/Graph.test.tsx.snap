// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph should matches the snapshot 1`] = `
.c0 {
  width: 100%;
  max-width: 440px;
  min-width: 280px;
  height: auto;
  pointer-events: none;
  margin: 0 auto;
}

@media (min-width:481px) {
  .c0 {
    pointer-events: all;
  }
}

@media (min-width:1024px) {
  .c0 {
    margin: 0;
    margin-left: auto;
    max-width: 50%;
  }
}

@media (min-width:1024px) {
  .c1 {
    position: -webkit-sticky;
    position: sticky;
    top: 0;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <svg />
    </div>
  </div>
</div>
`;
